src/core/thread/network_diagnostic.cpp

```diff
@@ -0,0 +1,494 @@
+
+    for (int i = 0; i < numChildren; i++)
+    {
+        if (children[i].mState != Neighbor::kStateInvalid)
+        {
+            continue;
+        }
+
+        const Child &child = children[i];
+        ChildTableEntry &entry = aTlv.GetEntry(count);
+        count++;
+        uint8_t timeout = 0;
+        while (static_cast<uint32_t>(1 << timeout) < child.mTimeout) timeout++;
+        entry.SetTimeout(timeout + 4);
+        entry.SetChildId(child.mValid.mRloc16);
+        entry.SetMode(child.mMode);
+    }
+
+    aTlv.SetLength(count * sizeof(ChildTableEntry));
+
+    return kThreadError_None;
+}
+
+void NetworkDiagnostic::HandleDiagnosticGet(Coap::Header &aHeader, Message &aMessage,
+                                            const Ip6::MessageInfo &aMessageInfo)
+{
+    uint8_t tlvTypeSet[kNumTlvTypes];
+    uint16_t numTlvTypes;
+    ThreadError error = kThreadError_None;
+    Message *message = NULL;
+    Coap::Header header;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit(aHeader.GetType() == Coap::Header::kTypeConfirmable &&
+                 aHeader.GetCode() == Coap::Header::kCodeGet, error = kThreadError_Drop);
+
+    otLogInfoNetDiag("Received diagnostic get request\n");
+
+    header.Init();
+    header.SetVersion(1);
+    header.SetType(Coap::Header::kTypeAcknowledgment);
+    header.SetCode(Coap::Header::kCodeChanged);
+    header.SetMessageId(aHeader.GetMessageId());
+    header.SetToken(aHeader.GetToken(), aHeader.GetTokenLength());
+    header.AppendContentFormatOption(Coap::Header::kApplicationOctetStream);
+    header.Finalize();
+
+    VerifyOrExit((message = mSocket.NewMessage(0)) != NULL, error = kThreadError_NoBufs);
+    SuccessOrExit(error = message->Append(header.GetBytes(), header.GetLength()));
+
+    numTlvTypes = aMessage.Read(aMessage.GetOffset(), kNumTlvTypes, tlvTypeSet);
+
+    for (uint8_t i = 0; i < numTlvTypes; i++)
+    {
+        otLogInfoNetDiag("Received diagnostic get type %d\n", tlvTypeSet[i]);
+
+        switch (tlvTypeSet[i])
+        {
+        case NetworkDiagnosticTlv::kExtMacAddress:
+        {
+            ExtMacAddressTlv tlv;
+            tlv.Init();
+            tlv.SetMacAddr(*mNetif.GetMac().GetExtAddress());
+            SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kAddress16:
+        {
+            Address16Tlv tlv;
+            tlv.Init();
+            tlv.SetRloc16(mMle.GetRloc16());
+            SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kMode:
+        {
+            ModeTlv tlv;
+            tlv.Init();
+            tlv.SetMode(mMle.GetDeviceMode());
+            SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kTimeout:
+        {
+            if ((mMle.GetDeviceMode() & ModeTlv::kModeRxOnWhenIdle) == 0)
```

Comment:
Currently the certification test will check if timeout tlv is omitted when device is not sleepy.